google-cloud-pubsub/lib/google/cloud/pubsub/dead_letter_policy.rb

```diff
@@ -0,0 +1,112 @@
+      #   `service-\\{project_number}@gcp-sa-pubsub.iam.gserviceaccount.com`) must have permission to Publish() to this
+      #   topic.
+      #
+      #   The operation will fail if the topic does not exist. Users should ensure that there is a subscription attached
+      #   to this topic since messages published to a topic with no subscriptions are lost.
+      # @attr [Numeric] max_delivery_attempts The maximum number of delivery attempts for any message in the
+      #   subscription's dead letter policy. Dead lettering is done on a best effort basis. The same message might be
+      #   dead lettered multiple times. The value must be between 5 and 100.
+      #
+      #   The number of delivery attempts is defined as 1 + (the sum of number of NACKs and number of times the
+      #   acknowledgement deadline has been exceeded for the message). A NACK is any call to ModifyAckDeadline with a 0
+      #   deadline. Note that client libraries may automatically extend ack_deadlines.
+      #
+      #   This field will be honored on a best effort basis.
+      #
+      # @example
+      #   require "google/cloud/pubsub"
+      #
+      #   pubsub = Google::Cloud::PubSub.new
+      #
+      #   sub = pubsub.subscription "my-topic-sub"
+      #   dead_letter_topic = pubsub.topic "my-dead-letter-topic", skip_lookup: true
+      #   sub.dead_letter_policy = Google::Cloud::PubSub::DeadLetterPolicy.new(
+      #     dead_letter_topic:     dead_letter_topic,
+      #     max_delivery_attempts: 20
+      #   )
+      #
+      #   sub.dead_letter_policy.dead_letter_topic.name #=> "projects/my-project/topics/my-dead-letter-topic"
+      #   sub.dead_letter_policy.max_delivery_attempts #=> 10
+      #
+      class DeadLetterPolicy
+        attr_reader :dead_letter_topic, :max_delivery_attempts
+
+        ##
+        # Creates a new, immutable DeadLetterPolicy value object.
+        #
+        # @attr [Topic, nil] dead_letter_topic The topic to which dead letter messages for the subscription should be
+        #   published. Dead lettering is done on a best effort basis. The same message might be dead lettered multiple
+        #   times. The Cloud Pub/Sub service account associated with the enclosing subscription's parent project (i.e.,
+        #   `service-\\{project_number}@gcp-sa-pubsub.iam.gserviceaccount.com`) must have permission to Publish() to
+        #   this topic.
+        #
+        #   The operation will fail if the topic does not exist. Users should ensure that there is a subscription
+        #   attached to this topic since messages published to a topic with no subscriptions are lost.
+        # @attr [Integer, nil] max_delivery_attempts The maximum number of delivery attempts for any message in the
+        #   subscription's dead letter policy. Dead lettering is done on a best effort basis. The same message might be
+        #   dead lettered multiple times. The value must be between 5 and 100.
```

Comment:
What would happen if the user specifies a value out of range? Should we have any local range checking, or are we depending on a server-sent 4xx error for out of range values?